components/brave_shields/browser/local_data_files_service.cc

```diff
@@ -0,0 +1,78 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_shields/browser/local_data_files_service.h"
+
+#include <algorithm>
+#include <utility>
+
+#include "base/base_paths.h"
+#include "base/bind.h"
+#include "base/logging.h"
+#include "base/macros.h"
+#include "base/memory/ptr_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/threading/thread_restrictions.h"
+#include "brave/browser/brave_browser_process_impl.h"
+#include "brave/components/brave_shields/browser/base_local_data_files_observer.h"
+
+namespace brave_shields {
+
+std::string LocalDataFilesService::g_local_data_files_component_id_(
+  kLocalDataFilesComponentId);
+std::string LocalDataFilesService::g_local_data_files_component_base64_public_key_(
+  kLocalDataFilesComponentBase64PublicKey);
+
+LocalDataFilesService::LocalDataFilesService()
+  : initialized_(false),
+    observers_already_called_(false),
+    weak_factory_(this) {
+  DETACH_FROM_SEQUENCE(sequence_checker_);
+}
+
+LocalDataFilesService::~LocalDataFilesService() { }
+
+bool LocalDataFilesService::Start() {
+  if (initialized_)
+    return true;
+  Register(kLocalDataFilesComponentName,
+           g_local_data_files_component_id_,
+           g_local_data_files_component_base64_public_key_);
+  initialized_ = true;
+  return true;
+}
+
+void LocalDataFilesService::AddObserver(BaseLocalDataFilesObserver* observer) {
+  if (observers_already_called_) {
```

Comment:
Is this to guard against the case where a client calls `AddObserver` after `Start` has been called? If so, maybe it would be better to just disallow calling `AddObserver` after `Start` (via an assertion or whatever).

Reply:
Good idea. I was originally concerned because the listening services are created on demand e.g. inside `BraveBrowserProcessImpl::tracking_protection_service()` so I wasn't sure they would be created in time, but if it's a problem we can do an `EnsureServicesAreCreated` kind of thing before calling `Start()` on the local data files component.